File: backend/seed_data/generate_seeds.py
"""Generate synthetic patient data for testing"""

import csv
import random
from datetime import datetime, timedelta
from pathlib import Path


# Inline cipher functions (no app dependency needed for generation)
def encrypt_medication(plain_text: str, age: int) -> str:
    shift = age % 26
    result = []
    for char in plain_text.upper():
        if char.isalpha():
            pos = ord(char) - ord("A")
            new_pos = (pos + shift) % 26
            result.append(chr(ord("A") + new_pos))
        else:
            result.append(char)
    return "".join(result)


def encode_telemetry(bpm: int, spo2: int) -> str:
    bpm_bytes = bpm.to_bytes(2, byteorder="big")
    spo2_bytes = spo2.to_bytes(2, byteorder="big")
    return (bpm_bytes + spo2_bytes).hex()


BASE_DIR = Path(__file__).parent
PATIENTS = 20
COLLIDING_IDS = 5

MEDICATIONS = [
    "ASPIRIN",
    "INSULIN",
    "MORPHINE",
    "WARFARIN",
    "LISINOPRIL",
    "METFORMIN",
    "ATORVASTATIN",
    "AMLODIPINE",
    "OMEPRAZOLE",
    "GABAPENTIN",
]

WARDS = ["ICU-1", "ICU-2", "ICU-3", "Ward-4", "Ward-5", "ER"]

FIRST_NAMES = [
    "John",
    "Jane",
    "Robert",
    "Maria",
    "David",
    "Sarah",
    "Michael",
    "Emily",
    "James",
    "Linda",
    "William",
    "Patricia",
    "Richard",
    "Barbara",
    "Joseph",
    "Elizabeth",
    "Thomas",
    "Susan",
    "Charles",
    "Jessica",
]
LAST_NAMES = [
    "Smith",
    "Johnson",
    "Williams",
    "Brown",
    "Jones",
    "Garcia",
    "Miller",
    "Davis",
    "Rodriguez",
    "Martinez",
    "Hernandez",
    "Lopez",
    "Gonzalez",
    "Wilson",
    "Anderson",
    "Thomas",
    "Taylor",
    "Moore",
    "Jackson",
    "Martin",
]


def generate_patients():
    """Generate patient demographics"""
    patients = []
    used_names = set()

    for i in range(PATIENTS):
        if i < COLLIDING_IDS * 2:
            raw_id = f"P{(i // 2) + 1:05d}"
        else:
            raw_id = f"P{i + 1:05d}"

        age = random.randint(18, 85)

        while True:
            name = f"{random.choice(FIRST_NAMES)} {random.choice(LAST_NAMES)}"
            if name not in used_names:
                used_names.add(name)
                break

        name_cipher = name.upper().replace(" ", "")

        patients.append(
            {
                "patient_raw_id": raw_id,
                "name_cipher": name_cipher,
                "age": age,
                "ward_code": random.choice(WARDS),
            }
        )

    with open(BASE_DIR / "patient_demographics.csv", "w", newline="") as f:
        writer = csv.DictWriter(
            f, fieldnames=["patient_raw_id", "name_cipher", "age", "ward_code"]
        )
        writer.writeheader()
        writer.writerows(patients)

    return patients


def generate_telemetry(patients):
    """Generate telemetry logs - 1000 samples per patient over 24 hours"""
    telemetry = []
    start_time = datetime.now() - timedelta(hours=24)

    for patient in patients:
        patient_idx = int(patient["patient_raw_id"][1:]) - 1
        if patient_idx < COLLIDING_IDS * 2:
            if patient_idx % 2 == 0:
                base_bpm = 70  # Even parity patient
            else:
                base_bpm = 75  # Odd parity patient
        else:
            base_bpm = random.randint(60, 90)

        for j in range(1000):
            timestamp = start_time + timedelta(minutes=j * 1.44)

            bpm = base_bpm + random.randint(-10, 10)
            spo2 = random.randint(94, 100)

            # 5% corrupted samples
            if random.random() < 0.05:
                if random.random() < 0.5:
                    hex_payload = "DEADBEEF"
                else:
                    bpm = random.randint(250, 300)
                    hex_payload = encode_telemetry(bpm, spo2)
            else:
                hex_payload = encode_telemetry(bpm, spo2)

            telemetry.append(
                {
                    "patient_raw_id": patient["patient_raw_id"],
                    "timestamp": timestamp.isoformat(),
                    "hex_payload": hex_payload,
                    "source_device": f"MONITOR_{random.randint(1, 10)}",
                }
            )

    with open(BASE_DIR / "telemetry_logs.csv", "w", newline="") as f:
        writer = csv.DictWriter(
            f,
            fieldnames=["patient_raw_id", "timestamp", "hex_payload", "source_device"],
        )
        writer.writeheader()
        writer.writerows(telemetry)

    return telemetry


def generate_prescriptions(patients):
    """Generate prescriptions with age-encrypted medication names"""
    prescriptions = []

    for patient in patients:
        num_meds = random.randint(3, 8)

        for _ in range(num_meds):
            med_name = random.choice(MEDICATIONS)
            age = patient["age"]
            med_cipher = encrypt_medication(med_name, age)

            prescriptions.append(
                {
                    "patient_raw_id": patient["patient_raw_id"],
                    "timestamp": (
                        datetime.now() - timedelta(hours=random.randint(1, 72))
                    ).isoformat(),
                    "age": age,
                    "med_cipher_text": med_cipher,
                    "dosage": f"{random.choice([2, 5, 10, 25, 50, 100])}mg",
                    "route": random.choice(["PO", "IV", "SC", "IM"]),
                }
            )

    with open(BASE_DIR / "prescription_audit.csv", "w", newline="") as f:
        writer = csv.DictWriter(
            f,
            fieldnames=[
                "patient_raw_id",
                "timestamp",
                "age",
                "med_cipher_text",
                "dosage",
                "route",
            ],
        )
        writer.writeheader()
        writer.writerows(prescriptions)

    return prescriptions


if __name__ == "__main__":
    print("Generating seed data...")

    patients = generate_patients()
    print(f"  Generated {len(patients)} patients")

    telemetry = generate_telemetry(patients)
    print(f"  Generated {len(telemetry)} telemetry samples")

    prescriptions = generate_prescriptions(patients)
    print(f"  Generated {len(prescriptions)} prescriptions")

    print("\nSeed data generated successfully!")
    print("  - patient_demographics.csv")
    print("  - telemetry_logs.csv")
    print("  - prescription_audit.csv")

File: backend/seed_data/generate_seeds.mjs
import { writeFileSync } from "fs";
import { join, dirname } from "path";
import { fileURLToPath } from "url";

const __dirname = dirname(fileURLToPath(import.meta.url));

function encryptMedication(plainText, age) {
  const shift = age % 26;
  return plainText
    .toUpperCase()
    .split("")
    .map((ch) => {
      if (/[A-Z]/.test(ch)) {
        const pos = ch.charCodeAt(0) - 65;
        return String.fromCharCode(65 + ((pos + shift) % 26));
      }
      return ch;
    })
    .join("");
}

function encodeTelemetry(bpm, spo2) {
  const bpmHex = bpm.toString(16).padStart(4, "0");
  const spo2Hex = spo2.toString(16).padStart(4, "0");
  return bpmHex + spo2Hex;
}

const PATIENTS = 20;
const COLLIDING_IDS = 5;

const MEDICATIONS = [
  "ASPIRIN", "INSULIN", "MORPHINE", "WARFARIN", "LISINOPRIL",
  "METFORMIN", "ATORVASTATIN", "AMLODIPINE", "OMEPRAZOLE", "GABAPENTIN",
];

const WARDS = ["ICU-1", "ICU-2", "ICU-3", "Ward-4", "Ward-5", "ER"];

const FIRST_NAMES = [
  "John","Jane","Robert","Maria","David","Sarah","Michael","Emily",
  "James","Linda","William","Patricia","Richard","Barbara","Joseph",
  "Elizabeth","Thomas","Susan","Charles","Jessica",
];
const LAST_NAMES = [
  "Smith","Johnson","Williams","Brown","Jones","Garcia","Miller",
  "Davis","Rodriguez","Martinez","Hernandez","Lopez","Gonzalez",
  "Wilson","Anderson","Thomas","Taylor","Moore","Jackson","Martin",
];

function randInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}
function randChoice(arr) {
  return arr[randInt(0, arr.length - 1)];
}

// Generate patients
const patients = [];
const usedNames = new Set();

for (let i = 0; i < PATIENTS; i++) {
  let rawId;
  if (i < COLLIDING_IDS * 2) {
    rawId = `P${String(Math.floor(i / 2) + 1).padStart(5, "0")}`;
  } else {
    rawId = `P${String(i + 1).padStart(5, "0")}`;
  }
  const age = randInt(18, 85);
  let name;
  do {
    name = `${randChoice(FIRST_NAMES)} ${randChoice(LAST_NAMES)}`;
  } while (usedNames.has(name));
  usedNames.add(name);
  const nameCipher = name.toUpperCase().replace(/ /g, "");
  patients.push({ patient_raw_id: rawId, name_cipher: nameCipher, age, ward_code: randChoice(WARDS) });
}

// Write patient_demographics.csv
let csv = "patient_raw_id,name_cipher,age,ward_code\n";
for (const p of patients) {
  csv += `${p.patient_raw_id},${p.name_cipher},${p.age},${p.ward_code}\n`;
}
writeFileSync(join(__dirname, "patient_demographics.csv"), csv);
console.log(`  Generated ${patients.length} patients`);

// Generate telemetry (1000 samples per patient)
const telemetryLines = ["patient_raw_id,timestamp,hex_payload,source_device"];
const startTime = new Date(Date.now() - 24 * 60 * 60 * 1000);

for (const patient of patients) {
  const patientIdx = parseInt(patient.patient_raw_id.slice(1)) - 1;
  let baseBpm;
  if (patientIdx < COLLIDING_IDS * 2) {
    baseBpm = patientIdx % 2 === 0 ? 70 : 75;
  } else {
    baseBpm = randInt(60, 90);
  }

  for (let j = 0; j < 1000; j++) {
    const ts = new Date(startTime.getTime() + j * 1.44 * 60000);
    let bpm = baseBpm + randInt(-10, 10);
    const spo2 = randInt(94, 100);
    let hexPayload;

    if (Math.random() < 0.05) {
      if (Math.random() < 0.5) {
        hexPayload = "DEADBEEF";
      } else {
        bpm = randInt(250, 300);
        hexPayload = encodeTelemetry(bpm, spo2);
      }
    } else {
      hexPayload = encodeTelemetry(bpm, spo2);
    }

    telemetryLines.push(
      `${patient.patient_raw_id},${ts.toISOString()},${hexPayload},MONITOR_${randInt(1, 10)}`
    );
  }
}
writeFileSync(join(__dirname, "telemetry_logs.csv"), telemetryLines.join("\n"));
console.log(`  Generated ${telemetryLines.length - 1} telemetry samples`);

// Generate prescriptions
const prescriptionLines = ["patient_raw_id,timestamp,age,med_cipher_text,dosage,route"];
for (const patient of patients) {
  const numMeds = randInt(3, 8);
  for (let m = 0; m < numMeds; m++) {
    const medName = randChoice(MEDICATIONS);
    const medCipher = encryptMedication(medName, patient.age);
    const ts = new Date(Date.now() - randInt(1, 72) * 3600000);
    const dosage = `${randChoice([2, 5, 10, 25, 50, 100])}mg`;
    const route = randChoice(["PO", "IV", "SC", "IM"]);
    prescriptionLines.push(
      `${patient.patient_raw_id},${ts.toISOString()},${patient.age},${medCipher},${dosage},${route}`
    );
  }
}
writeFileSync(join(__dirname, "prescription_audit.csv"), prescriptionLines.join("\n"));
console.log(`  Generated ${prescriptionLines.length - 1} prescriptions`);

console.log("\nSeed data generated successfully!");
console.log("  - patient_demographics.csv");
console.log("  - telemetry_logs.csv");
console.log("  - prescription_audit.csv");
